stan/math/prim/functor/reduce_sum.hpp

```diff
@@ -0,0 +1,227 @@
+#define STAN_MATH_PRIM_SCAL_FUNCTOR_REDUCE_SUM_HPP
+
+#include <stan/math/prim/meta.hpp>
+
+#include <tbb/task_arena.h>
+#include <tbb/parallel_reduce.h>
+#include <tbb/blocked_range.h>
+
+#include <tuple>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+namespace internal {
+
+/**
+ * reduce_sum_impl implementation for any autodiff type.
+ *
+ * @tparam ReduceFunction Type of reducer function
+ * @tparam ReturnType An arithmetic type
+ * @tparam Vec Type of sliced argument
+ * @tparam Args Types of shared arguments
+ */
+template <typename ReduceFunction, typename Enable, typename ReturnType,
+          typename Vec, typename... Args>
+struct reduce_sum_impl {
+  /**
+   * Call an instance of the function `ReduceFunction` on every element
+   *   of an input sequence and sum these terms.
+   *
+   * This specialization is not parallelized and works for any autodiff types.
+   *
+   * An instance, f, of `ReduceFunction` should have the signature:
+   *   T f(int start, int end, Vec&& vmapped_subset, std::ostream* msgs,
+   * Args&&... args)
+   *
+   * `ReduceFunction` must be default constructible without any arguments
+   *
+   * Each call to `ReduceFunction` is responsible for computing the
+   *   start through end - 1 terms of the overall sum. All args are passed
+   *   from this function through to the `ReduceFunction` instances.
+   *   However, only elements start through end - 1 of the vmapped argument are
+   *   passed to the `ReduceFunction` instances (as the `vmapped_subset`
+   * argument).
+   *
+   * @param vmapped Sliced arguments used only in some sum terms
+   * @param grainsize Suggested grainsize for tbb
+   * @param[in, out] msgs The print stream for warning messages
+   * @param args Shared arguments used in every sum term
+   * @return Summation of all terms
+   */
+  return_type_t<Vec, Args...> operator()(Vec&& vmapped, std::size_t grainsize,
+                                         std::ostream* msgs,
+                                         Args&&... args) const {
+    const std::size_t num_jobs = vmapped.size();
+
+    if (num_jobs == 0) {
+      return 0.0;
+    }
+
+    return ReduceFunction()(0, vmapped.size() - 1, std::forward<Vec>(vmapped),
+                            msgs, std::forward<Args>(args)...);
+  }
+};
+
+/**
+ * Specialization of reduce_sum_impl for arithmetic types
+ *
+ * @tparam ReduceFunction Type of reducer function
+ * @tparam ReturnType An arithmetic type
+ * @tparam Vec Type of sliced argument
+ * @tparam Args Types of shared arguments
+ */
+template <typename ReduceFunction, typename ReturnType, typename Vec,
+          typename... Args>
+struct reduce_sum_impl<ReduceFunction, require_arithmetic_t<ReturnType>,
+                       ReturnType, Vec, Args...> {
+  /**
+   * Internal object meeting the Imperative form requirements of
+   * `tbb::parallel_reduce`
+   *
+   * @note see link [here](https://tinyurl.com/vp7xw2t) for requirements.
```

Comment:
@bbbales2 check out the link here

> The parallel_reduce template has two forms. The functional form is designed to be easy to use in conjunction with lambda expressions. The imperative form is designed to minimize copying of data.

So the answer to "can this be a function" is probably yes however they say the imperative style does less copying. We could always try the functional style and benchmark the diff. idk anything about tbb's internals here so can't say whether the performance would be noticable

Reply:
I went with the imperative form for the promised additional efficiency... I don't need to benchmark this for my taste given that we do what is supposed to be faster and it's fine to me style wise to go with a struct.

Reply:
Aaah, I meant reduce_sum_impl, not recursive_reducer.